### DISCUSSION TITLE
What is the approved way of using `meson test` with meson-python?

### DISCUSSION THREAD

**Discussion by igo95862** [mesonbuild/meson-python#447] [Q&A]
I already use `meson test` in [my project](https://github.com/igo95862/bubblejail/tree/master/test) to run python unittests. However, I do it a little bit backwards. 

In the test folder every python test file has a `if __name__ == "__main__"` guard that runs the `unittest.main()` if the file is main script file. In meson I add a test for each file and run it with the python installation.

```meson
unittests = files(
    'test_helper.py',
    'test_profiles.py',
    'test_service_info.py',
    'test_auto_completion.py',
    'test_full_run.py',
)

foreach unittest : unittests
    test(
        fs.stem(unittest),
        py_installation,
        args : [unittest],
        env : test_env,
    )
endforeach
```

The magick that sets up the python code is the `test_env` environment.

```meson
test_env.set('PYTHONPATH', meson.current_build_dir())
```

The issue is I am not actually using `install_sources` to add the python files. Instead I use a `custom_target` with my own bytecode compiler. I was looking in to switching to `install_sources` but it seems to be impossible to replicate my current setup.
`install_sources` does not actually copy the python files to the build directory which makes it impossible to have a single `PYTHONPATH` to load the python package for the unit tests. Not sure how well using  multiple paths in `PYTHONPATH` where one path is the source directory and the other one is build directory would work when there are both python files and extension modules.

**Comment by rgommers** [comment]
Hi @igo95862, thanks for your question. I'm not actually sure about my answer here, because there's multiple unusual things in your setup. A few thoughts:
- `meson-python` is only a build backend, meaning the only way to interact with it is through a build front-end like `pip` or `build`. At that point you get either an artifact (sdist and/or wheel) or an editable install. None of those are expected to work with `meson test`
- Use of `PYTHONPATH` manipulation is considered an anti-pattern; it's sometimes useful, but should in general not be a primary or only way to achieve your goal.
- For test files, there's usually a decision to make about whether or not to install them as part of your package. I'm not sure what you want there? If you do install them, can you run them with `pytest --pyargs bubblejail`?
- You don't have a `pyproject.toml` file in your repo, so you aren't using meson-python, right?
- Using `meson test` is unusual for a Python project, I'd expect it more for a mixed-language project with non-Python installable components. At that point, for development workflow you don't need meson-python. The two don't mix well.

**Reply by igo95862** [reply]
> * `meson-python` is only a build backend, meaning the only way to interact with it is through a build front-end like `pip` or `build`. At that point you get either an artifact (sdist and/or wheel) or an editable install. None of those are expected to work with `meson test`

Maybe it is more of the question to the `meson` project itself and its `python` module.

The `--editable` pip mode might be a replacement since it recompiles the extensions on source change. The unit tests then could be run in a venv `./venv/bin/python -m unittest`. 

Not sure how to integrate the meson's `run_target` though.

**Comment by Mohataseem89** [comment]
`meson-python` is not designed to integrate with `meson test` directly. It is a build backend only so its output is an installable Python artifact (sdist/wheel) or an editable install. Once you reach that stage, Meson is no longer the test runner Python tooling is

The approved/supported workflows are:
1. Use an editable install and run tests with Python tooling
This is the closest replacement for what you are doing with `PYTHONPATH` today. In a virtualenv:
```
pip install -e .
pytest
# or
python -m unittest
```
Editable mode automatically rebuilds extension modules on source change so you do n0t need to point `PYTHONPATH` at build directories

2. Treat Meson as the build system, and test the installed artifact
If you want to test the wheel as Meson produces it:
```
pip wheel .
pip install dist/*.whl
pytest --pyargs <package>
```
this validates the actual installation layout rather than a stitched-together PYTHONPATH environment.


Why `meson test` is not recommended
`meson test` works well when:
the project is C/C++/Fortran first
you want to test executables in the builddir

But for a Python-installable package, Meson intentionally does not stage Python sources in the build dir the way it does for native code. It delegates packaging to the Python ecosystem. That is why `install_sources()` does not mirror the tree into the build dir, and why a single unified `PYTHONPATH` isn’t expected to work.

why `PYTHONPATH` hacks are discouraged
it bypasses packaging semantics
it hides import errors
mixed import roots (source + .so in builddir) can mask ABI issues
Editable mode is the correct “development import” story.

what about `run_target()`?
`run_target()` is a Meson-side concept. Once the project is in editable mode, you can wrap that logic in a standard script and invoke it through normal Python entry points. Meson isn’t meant to orchestrate Python-module execution after the build backend hands control to pip.




If your end goal is Python unit-testing, the supported answer is:
Build/install with `meson-python`, then run tests via Python (pytest/unittest), preferably using `pip install -e` . during development.
`meson test` is not part of the meson-python workflow and there is no “approved” mechanism to make `meson test` act like a python test runner

**Comment by Mohataseem89** [comment]
There is not an “approved” or built-in way to combine `meson test` with meson-python. meson-python is a PEP 517 build backend only, so once you invoke it through `pip` or `build` you end up with either:
  - a wheel / sdist, or
  - an editable install via `pip install -e `.

At that point you are expected to use Python’s normal testing workflows, not `meson test`. The two layers are intentionally separated.

The supported approaches today are:
1. Use editable installs for development
```
python -m venv venv
source venv/bin/activate
pip install -e .
pytest      # or python -m unittest
```
Editable mode will rebuild extension modules automatically, which replaces the use of a custom `PYTHONPATH` pointed at the Meson builddir.

2. Test the built artifact
```
pip wheel .
pip install dist/*.whl
pytest --pyargs <your package>
```
That exercises the actual installation layout rather than tests against files staged in the builddir.

Why `meson test` isn’t recommended here:
  - Meson’s testing model assumes artifacts live in the builddir
  - meson-python does not stage Python sources into the builddir (by design)
  - PYTHONPATH tricks against a mixed set of import roots easily lead to masking issues or ABI mismatches
  
So if the final aim is Python unit tests, the supported workflow is:
  - Build/install with meson-python and run tests via Python tooling (pytest/unittest), usually with an editable install during development.
